### PR TITLE
add support for --new, --copy, --edit and --show

### PR DESCRIPTION
**Pull request by boegel**
* `--new`: create new easyconfig file based on specified information
* `--copy`: copy easyconfig file to specified location (or current directory)
* `--edit`: open easyconfig file with editor command specified via `--editor-command-template`
* `--show`: print contents of easyconfig file

This PR is mostly about adding the support for `--new`. The idea here is that you can basically just throw stuff at it that should be included in a new easyconfig file. The implementation uses a bunch of heuristics to figure out what the provided information likely represents. As such, it certainly won't be perfect in all possible situations, but I'm confident it will be very useful in practice.

The other options (`--copy`, `--edit` and `--show`) are intended to be used together with either `--new` or `--search`, and have a pretty straightforward implementation.

Example usage (docs update coming up later):

```
$ eb --new example 1.2.3 foss/2018a ConfigureMake files:bin/example "This an example description" homepage=https://example.com https://files/example.com/example-1.2.3.tar.gz moduleclass=tools --show
== temporary log file in case of crash /tmp/eb-rHjhuK/easybuild-uKEQ4P.log
== Contents of easyconfig file ./example-1.2.3-foss-2018a.eb:

easyblock = 'ConfigureMake'

name = 'example'
version = '1.2.3'

homepage = {'https': '//example.com'}
description = "This an example description"

toolchain = {'name': 'foss', 'version': '2018a'}

source_urls = ['https://files/%(name)s.com']
sources = [SOURCE_TAR_GZ]


sanity_check_paths = {
    'files': ['bin/%(name)s'],
    'dirs': [],
}

moduleclass = 'tools'

== easyconfig file ./example-1.2.3-foss-2018a.eb created!
```

### CODE REVIEW

File: easybuild/framework/easyconfig/tools.py [hunk 1]
```diff
@@ -699,3 +701,187 @@ def avail_easyblocks():
...
+        parts = item.split(sep)
+        return (parts[0], sep.join(parts[1:]))
+
+    param, value = None, None
+
+    # determine list of names of known easyblocks, so we can descriminate an easyblock name
+    easyblock_names = [e['class'] for e in avail_easyblocks().values()]
+
+    # regular expression to recognise a version
+    version_regex = re.compile('^[0-9][0-9.-]')
```
[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/framework/easyconfig/tools.py]: The assumption here is that a version always starts with a digit, and is followed by either another digit or a `.` or ` -`.

Is that too restrictive? I can loosen it up, but I think that would fit most versions out there?

You can always use `eb --new-pr example version=1.2.3`, that bypasses this heuristic entirely...

File: easybuild/framework/easyconfig/tools.py [hunk 2]
```diff
@@ -699,3 +701,187 @@ def avail_easyblocks():
                                        easyblock_mod_name, easyblocks[easyblock_mod_name]['loc'], path)
 
     return easyblocks
+
+
+def parse_param_value(string):
```
[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/framework/easyconfig/tools.py]: Very good point, thanks, I'll look into that.

**Comment by boegel** [comment]
@damianam W.r.t. your concern about flexibility: you can always bypass the heuristics by always passing `<param>=<value>` style arguments, which get preference over anything else.

For anything that doesn't match this pattern, there are a number of very reasonable heuristics.

One thing I should also clarify is that `eb --new` only generates an easyconfig file, it does *not* start the actual build afterwards. So, you're supposed to check the generated easyconfig file before using it, not use it blindly...

File: easybuild/framework/easyconfig/tools.py [hunk 3]
```diff
@@ -699,3 +701,187 @@ def avail_easyblocks():
...
+
+        # if we encounter a dictionary with (only) 'files' and/or 'dirs' as key(s), it must be sanity_check_paths
+        if len(value) <= 2 and ('files' in value or 'dirs' in value):
+            param = 'sanity_check_paths'
+            for item_key, item_val in value.items():
+                if isinstance(item_val, basestring):
+                    value[item_key] = []
+                    if item_val:
+                        value[item_key].append(item_val)
+                elif isinstance(item_val, tuple):
```
[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/framework/easyconfig/tools.py]: No, because `;` is used as a separator between different `<key>:<value>` entries in a dict value. And `,` is the separator for tuples (see below).

`;` is also the separator for lists, but since anything that starts with `[a-z_]+:` is considered a dict value, it'll get split on `;` to separate different `<key>=<value>` entries, and hence you can never have a list-type value anymore... :)

Example: `files:bin/example,lib/libexample.a;dirs:include` gets parsed as:

```python
{
  'files': ['bin/example', 'lib/libexample.a'],
  'dirs': ['include'],
}
```

`;` can not be used as a separator between `bin/example` and `lib/libexample.a`

File: easybuild/main.py [hunk 4]
```diff
@@ -173,6 +173,73 @@ def run_contrib_style_checks(ecs, check_contrib, check_style):
     return check_contrib or check_style
 
 
+def handle_cat_copy_edit(filepaths, target=None, copy=None):
```
[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/main.py]: The result of `handle_cat_copy_edit` is only used to print a message mentioning which files have been copied.

If you only use `--edit`, it'll open the file for editing in place, without copying it.

So you can go `eb --search foo=1.2.3 --edit` (`--edit` & co work together with both the added `--new` and the existing `--search`).

File: easybuild/framework/easyconfig/tools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (damianam) [reply easybuild/framework/easyconfig/tools.py]: Well, maybe "hyperrestrictive" is a stretch. But we have a couple of cases where version is a single digit, or a digit followed by a dash (`-`). I guess we can consider those corner cases not considered in the heuristic?

File: test/framework/easyconfig.py [hunk 5]
```diff
@@ -2357,6 +2357,59 @@ def test_deprecated(self):
         error_pattern = r"easyconfig file '.*/test.eb' is marked as deprecated:\nthis is just a test\n\(see also"
         self.assertErrorRegex(EasyBuildError, error_pattern, EasyConfig, test_ec)
 
+    def test_parse_param_value(self):
+        """Tests for parse_param_value function."""
+        self.assertEqual(parse_param_value(''), (None, ''))
+        self.assertEqual(parse_param_value('test123'), (None, 'test123'))
+
+        self.assertEqual(parse_param_value("this is a description"), ('description', "this is a description"))
```
Comment (damianam) [inline test/framework/easyconfig.py]: Have a test where "description" is not part of the string?

File: test/framework/easyconfig.py [hunk 6]
```diff
@@ -2357,6 +2357,59 @@ def test_deprecated(self):
...
+
+        self.assertEqual(parse_param_value('1.2.3'), ('version', '1.2.3'))
```
Comment (damianam) [inline test/framework/easyconfig.py]: Test different versions? `1.0`, `22.0`, `1.2a`....

File: test/framework/easyconfig.py [hunk 7]
```diff
@@ -2357,6 +2357,59 @@ def test_deprecated(self):
...
+
+        # names of source tarballs are recognized
+        for ext in ['tar.gz', 'gz', 'tar.bz2', 'bz2', 'zip']:
+            self.assertEqual(parse_param_value('toy-0.0.' + ext), ('sources', 'toy-0.0.' + ext))
```
Comment (damianam) [inline test/framework/easyconfig.py]: What if the file is `toy_0.0.zip` or `toy0.0.tar`?
